### PR TITLE
Add AutomatedFarfield

### PR DESCRIPTION
**Pull request by benflexcompute**
Done:
- [x] Refactor/fix the entity merging logic.
- [x] Move utility functions/classes out into util files.
- [x] Added append method to UniqueItemList to trigger validation when appending.
- [x] Added UniqueStringList for storing boundary names, The unique item list is buggy (do not know why) when serializing. Likely the serializing and deserializing is not self-consistent in this particular situation.
- [x] Added function to populate the list of boundaries a zone has with volume mesh metadata.
- [x] Added `AutomatedFarfield` and `GhostSurface` and unit test.

### CODE REVIEW

File: flow360/component/simulation/meshing_param/face_params.py [hunk 1]
```diff
@@ -31,12 +31,13 @@ class SurfaceRefinement(Flow360BaseModel):
         description="Local maximum edge length for surface cells."
     )
     # pylint: disable=no-member
-    curvature_resolution_angle: AngleType.Positive = pd.Field(
+    curvature_resolution_angle: Optional[AngleType.Positive] = pd.Field(
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/meshing_param/face_params.py:34]: is there any default? Maybe we can use 15 deg?

File: flow360/component/simulation/meshing_param/params.py [hunk 2]
```diff
@@ -74,6 +72,40 @@ class MeshingParams(Flow360BaseModel):
...
+    def _finalize_automated_farfield(cls, v) -> Self:
+        if v is None:
+            # User did not put anything in volume_zones so may not want to use volume meshing
+            return v
+
+        has_rotating_zone = False
+        for volume_zone in v:
+            if isinstance(volume_zone, RotationCylinder):
+                has_rotating_zone = True
+                break
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/meshing_param/params.py:90]:
```suggestion
        has_rotating_zone = any(isinstance(volume_zone, RotationCylinder) for volume_zone in v)
```

File: flow360/component/simulation/meshing_param/params.py [hunk 3]
```diff
@@ -74,6 +72,40 @@ class MeshingParams(Flow360BaseModel):
...
+    def _check_volume_zones_has_farfied(cls, v) -> Self:
+        if v is None:
+            # User did not put anything in volume_zones so may not want to use volume meshing
+            return v
+
+        has_farfield = False
+        for volume_zone in v:
+            if isinstance(volume_zone, AutomatedFarfield):
+                has_farfield = True
+                break
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/meshing_param/params.py:108]:
```suggestion
        has_farfield = any(isinstance(volume_zone, AutomatedFarfield) for volume_zone in v)
```

File: flow360/component/simulation/meshing_param/volume_params.py [hunk 4]
```diff
@@ -41,25 +56,99 @@ class AxisymmetricRefinement(Flow360BaseModel):
...
     https://www.notion.so/flexcompute/Python-model-design-document-
     78d442233fa944e6af8eed4de9541bb1?pvs=4#c2de0b822b844a12aa2c00349d1f68a3
+
+    - `enclosed_entities` is actually just a way of specifying the enclosing patches of a volume zone.
+    Therefore in the future when supporting arbitrary-axisymmetric shaped sliding interface, we may not need this
+    attribute at all. For example if the new class already has an entry to list all the enclosing patches.
+
     """
 
-    name: Optional[str] = pd.Field(None)
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/meshing_param/volume_params.py:57]: can we keep "name" in this PR and update all names all together when we have decision?

File: flow360/component/simulation/framework/param_utils.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    Note:
+    1. asset_entity_registry will be replacing/update the metadata-constructed registry of the asset when loading it.
+    """
+
+    asset_entity_registry: EntityRegistry = pd.Field(EntityRegistry(), frozen=True)
+    # pylint: disable=no-member
+    project_length_unit: Optional[LengthType.Positive] = pd.Field(None, frozen=True)
+
+
+def recursive_register_entity_list(model: Flow360BaseModel, registry: EntityRegistry) -> None:
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/param_utils.py:27]: Done

File: flow360/component/simulation/framework/param_utils.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        elif isinstance(field, list):
+            for item in field:
+                if isinstance(item, Flow360BaseModel):
+                    recursive_register_entity_list(item, registry)
+
+        elif isinstance(field, Flow360BaseModel):
+            recursive_register_entity_list(field, registry)
+
+
+def _recursive_update_zone_name_in_surface_with_metadata(
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/param_utils.py:64]: Done

File: flow360/component/simulation/meshing_param/face_params.py (hunk 1 above)
Comment (benflexcompute) [reply flow360/component/simulation/meshing_param/face_params.py:34]: Added 12 deg

File: flow360/component/simulation/meshing_param/params.py (hunk 2 above)
Comment (benflexcompute) [reply flow360/component/simulation/meshing_param/params.py:90]: Done

File: flow360/component/simulation/meshing_param/params.py (hunk 3 above)
Comment (benflexcompute) [reply flow360/component/simulation/meshing_param/params.py:108]: Done

File: flow360/component/simulation/meshing_param/volume_params.py (hunk 4 above)
Comment (benflexcompute) [reply flow360/component/simulation/meshing_param/volume_params.py:57]: Ok

File: flow360/component/simulation/framework/entity_registry.py [hunk 7]
```diff
@@ -60,7 +60,7 @@ def register(self, entity: EntityBase):
         # pylint: disable=unsubscriptable-object
         self.internal_registry[entity.entity_bucket].append(entity)
 
-    def get_all_entities_of_given_bucket(self, entity_bucket):
+    def get_all_entities_of_given_bucket(self, entity_bucket) -> list[EntityBase]:
```
[the remark this answers is not in the document]
Comment (maciej-flexcompute) [reply flow360/component/simulation/framework/entity_registry.py]: the benefit is in more granular responsibility. With this design I can see that:
1. "registry" is responsible of containing "buckets" and giving access to buckets
2. "buckets" are responsible for holding "entities" and giving access to entities
When you want to add new functionality, you know who should be responsible for it so you know where do add code. On the other hand, `get_all_entities_of_given_bucket` is a very specialised function. It is not clear what is bucket, what is entity and what is relationship between them

**Review by maciej-flexcompute — APPROVED** [review]

File: flow360/component/simulation/framework/param_utils.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+from flow360.component.simulation.utils import _model_attribute_unlock
+
+
+class AssetCache(Flow360BaseModel):
+    """
+    Note:
+    1. asset_entity_registry will be replacing/update the metadata-constructed registry of the asset when loading it.
+    """
+
+    asset_entity_registry: EntityRegistry = pd.Field(EntityRegistry(), frozen=True)
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/framework/param_utils.py:22]: can we just call it `registry`?

File: flow360/component/simulation/meshing_param/params.py [hunk 9]
```diff
@@ -74,6 +72,34 @@ class MeshingParams(Flow360BaseModel):
...
+    volume_zones: Optional[List[Union[RotationCylinder, AutomatedFarfield]]] = pd.Field(
+        default=None, description="Creation of new volume zones."
     )
+
+    @pd.field_validator("volume_zones", mode="after")
+    @classmethod
+    def _finalize_automated_farfield(cls, v) -> Self:
+        if v is None:
+            # User did not put anything in volume_zones so may not want to use volume meshing
+            return v
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/meshing_param/params.py:84]: I would assume that if you don't want to use meshing, entire object "meshing" should be None. If meshing exists, I would assume it is indented to be used

File: flow360/component/simulation/meshing_param/params.py [hunk 10]
```diff
@@ -47,9 +48,6 @@ class MeshingParams(Flow360BaseModel):
     """
 
     # Volume **defaults**:
-    farfield: Optional[Literal["auto", "quasi-3d", "user-defined"]] = pd.Field(
-        default="auto", description="Type of farfield generation."
-    )
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/meshing_param/params.py:52]: how do we keep this behaviour: default="auto" ? I would prefer user does not need to do it explicitly especially on webUI?

File: flow360/component/simulation/framework/param_utils.py (hunk 8 above)
Comment (benflexcompute) [reply flow360/component/simulation/framework/param_utils.py:22]: you mean the asset_entity_registry or the EntityRegistry class?

File: flow360/component/simulation/meshing_param/params.py (hunk 9 above)
Comment (benflexcompute) [reply flow360/component/simulation/meshing_param/params.py:84]: Yes but if the user only wants the surface meshing then the volume_zones can be None.

File: flow360/component/simulation/meshing_param/params.py (hunk 10 above)
Comment (benflexcompute) [reply flow360/component/simulation/meshing_param/params.py:52]: We can change this in the get_default_param service?

File: flow360/component/simulation/framework/param_utils.py (hunk 8 above)
Comment (maciej-flexcompute) [reply flow360/component/simulation/framework/param_utils.py:22]: `asset_entity_registry`

Comment (benflexcompute) [reply flow360/component/simulation/framework/param_utils.py]: Done

**Pull request merged**